peps/pep-0728.rst

```diff
@@ -0,0 +1,632 @@
+PEP: 728
+Title: TypedDict with Typed Extra Fields
+Author: Zixuan James Li <p359101898@gmail.com>
+Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
+Discussions-To:
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Content-Type: text/x-rst
+Created: 12-Sep-2023
+Python-Version: 3.13
+Post-History:
+
+
+.. highlight:: rst
+
+Abstract
+========
+
+This PEP proposes a way to type extra fields for :class:`~typing.TypedDict` using
+a using a special ``__extra__: T`` field.
+This addresses the need to define a subset of fields that might appear in a
+``dict`` while permitting additional fields of a specified type, and the need to
+create a closed TypedDict type with ``__extra__: Never``.
+
+Motivation
+==========
+
+A TypedDict type can annotate the type of each known field from dictionaries.
+However, due to structural subtyping, the TypedDict can have extra fields that
+are not visible through its type. There is currently not a way to limit or
+restrict the type of fields that might be present in the TypedDict type's
+structural subtype.
+
+Defining a Closed TypedDict Type
+--------------------------------
+
+In practice, this prevents users from defining a closed TypedDict type
+when it is expected that the TypedDict type contains no additional fields.
+
+Due to the possible presence of extra fields, type checker cannot infer more
+precise return types for ``.items()`` and ``.values()`` on a TypedDict. This can
+also be resolved by
+`defining a closed TypedDict type <https://github.com/python/mypy/issues/7981>`__.
+
+Another possible use case for this is a sound way to
+`enable type narrowing <https://github.com/python/mypy/issues/9953>`__ with the
+``in`` check::
+
+    class Movie(TypedDict):
+        name: str
+        director: str
+    
+    class Book(TypedDict):
+        name: str
+        author: str
+
+    def fun(entry: Movie | Book) -> None:
+        if "author" in entry:
+            add_book(entry)  # Not OK
+
+Nothing prevents a ``dict`` that is structurally compatible with ``Movie`` to
+have the ``author`` key, and it would be correct for the type checker to reject
+the ``add_book`` call.
+
+Allowing Extra Fields of a Certain Type
+---------------------------------------
+
+For supporting API interfaces or legacy codebase where only a subset of possible
+keys are known, it would be useful to explicitly expect additional keys of
+certain value types.
+
+However, :pep:`589` is more restrictive on type checking the construction of a
```

Comment:
Nit: We should ideally frame this around the current typing spec, which derives from PEP 589 but can be changed independently.